src/eko/io/types.py

```diff
@@ -0,0 +1,162 @@
+    issubclass(RunningReference, DictLike)  # raise an error, since
+                                            # RunningReference is not a class
+
+Essentially classes can be used for type hints, but types are not all classes,
+especially when they involve generics.
+
+For this reason I prefer the less elegant dynamic generation, that seems to
+preserve type hints.
+
+"""
+import dataclasses
+import enum
+import typing
+from math import isnan
+from typing import Any, Dict
+
+from .dictlike import DictLike
+
+T = typing.TypeVar("T")
+Quantity = typing.TypeVar("Quantity", bound=typing.Union[int, float])
+
+
+def reference_running(quantity: typing.Type[Quantity]):
+    """Generate running quantities reference point classes.
+
+    The motivation for dynamic generation is provided in module docstring.
+
+    """
+
+    @dataclasses.dataclass
+    class ReferenceRunning(DictLike):
+        value: quantity
+        scale: float
+
+    return ReferenceRunning
+
+
+IntRef = reference_running(int)
+FloatRef = reference_running(float)
+
+Order = typing.Tuple[int, int]
+
+
+@dataclasses.dataclass
+class CouplingsRef(DictLike):
+    """Reference values for coupling constants."""
+
+    alphas: FloatRef
+    alphaem: FloatRef
+
+    def __post_init__(self):
+        """Validate couplings.
+
+        If they are both running, they have to be defined at the same scale.
+
+        Usually :attr:`alphaem` is not running, thus its scale is set to nan.
+
+        """
+        assert self.alphas.scale == self.alphaem.scale or isnan(self.alphaem.scale)
+
+    @property
+    def values(self):
+        """Collect only couplings values."""
+        return (self.alphas.value, self.alphaem.value)
+
+
+FlavorsNumber = int
+FlavorIndex = int
+IntrinsicFlavors = typing.List[FlavorIndex]
+
+
+def heavy_quark(quarkattr):
+    """Generate heavy quark properties container classes.
+
+    The motivation for dynamic generation is provided in module docstring.
+
+    """
+
+    @dataclasses.dataclass
+    class HeavyQuarks(DictLike):
+        """Access heavy quarks attributes by name."""
+
+        c: quarkattr
+        """Charm quark."""
+        b: quarkattr
+        """Bottom quark."""
+        t: quarkattr
+        """Top quark."""
+
+        def __getitem__(self, key: int):
+            """Allow access by index.
+
+            Consequently it allows iteration and containing check.
+
+            """
+            return getattr(self, "cbt"[key])
+
+    return HeavyQuarks
+
+
+QuarkMassRef = FloatRef
+HeavyQuarkMasses = heavy_quark(QuarkMassRef)
+MatchingScale = float
+MatchingScales = heavy_quark(MatchingScale)
+
+
+# TODO: upgrade all the following to StrEnum, new in py3.11
+# with that, it is possible to replace all non-alias right sides with calls to
+# enum.auto()
+
+
+class QuarkMassSchemes(enum.Enum):
+    """Scheme to define heavy quark masses."""
+
+    MSBAR = "msbar"
+    POLE = "pole"
+
+
+class EvolutionMethod(enum.Enum):
```

Comment:
I wonder whether we should lift this stuff up (and out of `io`) since these informations also need to be passed around in eko, e.g. here https://github.com/NNPDF/eko/blob/edf78a6a32bd159f78934c75e24ccb2d5f6acec9/src/eko/evolution_operator/__init__.py#L120 and we already found out that passing strings around in numba is slowing the thing down ... so, e.g., can we have both a string mapping for users and int mapping for numba?

if we want to do this, then not in this PR

Reply:
> I wonder whether we should lift this stuff up (and out of `io`) since these informations also need to be passed around in eko, e.g. here

Yes, this I considered as well. Actually, the first instance of the `types.py` file appeared top-level, not inside `eko.io`, but then I moved in since it was mostly related to/used in runcards.
Type hints are extremely useful, I'm decorating even tests because they are giving me autocompletion (and it saves me the time of doing manually the lookup for attributes and methods), so maybe we want really to have eko-wide types.

Reply:
> and we already found out that passing strings around in numba is slowing the thing down ... so, e.g., can we have both a string mapping for users and int mapping for numba?

I guess it's feasible, even though I'm not currently sure as the best way to implement it (a tuple value? some built-in? dataclass value?). However, we can reconsider later on.

Reply:
Now a dedicated issue #181, we're not going to face properly in this PR